Blog/how to make your first gem (#204)
rootstrap/blog: content/How to Make My First Gem - Timothy Peraza.md

> # **How To Make Your First Ruby Gem ( a step-by-step guide)**

Reviewer: ```suggestion
# **How To Make Your First Ruby Gem (a step-by-step guide)**
```

It could be argued, that most developers who have worked with ruby on rails, might be interested in developing a specific gem to solve problems for their fellow developers. 

> However, it's not uncommon for developers to not know where to begin, or where a gem is created in the first place. It's also not uncommon for a developer to feel like this would be a difficult and complicated task. For that reason, the aim of this article is to explain how to develop your first gem, and also take a look at the different components involved.  

Reviewer: ```suggestion
However, it's not uncommon for developers to not know where to begin, or to know where a gem is created in the first place. It's also not uncommon for a developer to feel like this would be a difficult and complicated task. For that reason, the aim of this article is to explain how to develop your first gem, and also take a look at the different components involved.  
```

<img src="images/rsgem_structure.png"
     alt="Blog Gem"
     width="500" height="250"
     align="middle"/>

It's important to mention that the gem rsgem is based on the command 'bundle gem', said command is the one that generates the directories and the files that we see. Now, we will mention the different components which are generated for the gem that we want to create.

## Bin folder

<img src="images/bin_folder.png"
     alt="Bin folder"
     width="500" height="250"
     align="middle"/>

In the Bin folder, we can find the console and setup files. The important thing to highlight here is the utility of console file where we can see the next code:

```ruby
#!/usr/bin/env ruby

require 'bundler/setup'
require 'blog_gem'

# You can add fixtures and/or initialization code here to make experimenting
# with your gem easier. You can also use a different console, if you like.

# (If you use this, don't forget to add pry to your Gemfile!)
# require "pry"
# Pry.start

require 'irb'
IRB.start(__FILE__)
```

> If we have to do test from the terminal with our gem, we only need the command bundle console 'require_name' to be able to access to the interactive ruby shell with each of our methods and dependecies of the gem. In our example of gem, we have just to do bundle console blog_gem.

Reviewer: ```suggestion
If we have to test from the terminal with our gem, we just need the command bundle console 'require_name', to access the interactive ruby shell with each of our methods, and dependencies of the gem. In our gem example, we just need to do bundle console blog_gem.
```

> # **How to make your first Ruby Gem (step-by-step)**

Reviewer: ```suggestion
# **How To Make Your First Ruby Gem ( a step-by-step guide)**
```

> Every developer, who has worked with ruby on rails and has felt pasion or intenrest for the open source world, has wanted or wants to do a contribution to this community of development with a gem which could solves a particular or generic problem or improve existing one, that appears in differents projects across the developer carrer. 

Reviewer 2: ```suggestion
Every developer, who has worked with ruby on rails and has felt passion or interest for the open source world, has wanted or wants to do a contribution to this community of development with a gem which could solve a particular or generic problem or improve an existing one, that appears in different projects across the developer carrer. 
```

Reviewer 3: Career instead of "carrer"

> Every developer who has worked with ruby on rails and has felt passion or interest for the open source world, has wanted or wants to do a contribution to this community of development with a gem which could solve a particular or generic problem or improve an existing one, that appears in different projects across the developer career.

Reviewer: ```suggestion
Every developer who has worked with ruby on rails has in some way wanted to contribute to this broad community of developers, with a gem that could solve a particular or generic problem across the spam of a developer's career.  
```

> However, sometimes happen that one doesn't know where to start or how is a gem made or other times one thinks or has the idea that to make a gem is a really hard or complecated task. For that reason, in this articule is searched to inquire how to make your first gem and to explain a bit the different component of its.

Reviewer 2: ```suggestion
However, sometimes it happens that one doesn't know where to start or how is a gem made in the first place, or other times one thinks or has the idea that to make a gem is a really hard or complicated task. For that reason, this article is aimed to inquire how to make your first gem and to explain a bit the different components of it.
```

> Here at Rootstrap, we are much interested to be reconized in the open source world. For that, from a time, we have been working in this topic, making libraries for diferents lenguajes and frameworks and generating visible with blogs and posts in social media. In particular, we built [rsgem](https://github.com/rootstrap/rsgem), a gem which generates a base project to make a gem in which one this article is based on.

Reviewer 2: ```suggestion
Here at Rootstrap, we are very interested in being reconized in the open source world. For that reason, we have been working in this topic for some time, making libraries for diferents languages and frameworks and making them visible with blogs and posts in social media. In particular, we built [rsgem](https://github.com/rootstrap/rsgem), a gem which generates a base project to make a gem, in which one this article is based on.
```

> ## Stargint the Gem

Reviewer 2: ```suggestion
## Starting the Gem
```
?

> Here at Rootstrap, we are very interested in being reconized in the open source world. For that reason, we have been working in this topic for some time, making libraries for diferents languages and frameworks and making them visible with blogs and posts in social media. In particular, we made [rsgem](https://github.com/rootstrap/rsgem), a gem which generates a base project to make a gem, in which one this article is based on.

Reviewer: ```suggestion
Here at Rootstrap, we are very interested in being a part of the open-source world. For that reason, we have been working on this topic for some time and making libraries for different languages/frameworks, and also making them visible publicly. For example, we made [rsgem](https://github.com/rootstrap/rsgem), a gem that allows you to generate a base project to make a gem, in which this article is based on.
```

> ## Starting the Gem

Reviewer: ```suggestion
## Starting The Gem
```

> The first step to start is to install the rsgem, this is able to do easily using the command gem install rsgem. After we have the gem installed, we have to run the command rsgem new name, with the name that we wante for our gem. In particular, in this blog, we run the command rsgem new blog_gem, how you see this will generate a project with the next structure

Reviewer 2: ```suggestion
The first step to start is to install the rsgem, this is able to do easily by running `$ gem install rsgem`. After we have the gem installed, we have to run `$ rsgem new [name]`, with the name that we want for our gem. In particular, in this blog, we will be running the command `$ rsgem new blog_gem`, and this will generate a project with the following structure:
```

> The first step to start is to install the rsgem, this is able to do easily using the command gem install rsgem. After we have the gem installed, we have to run the command rsgem new name, with the name that we wante for our gem. In particular, in this blog, we run the command rsgem new blog_gem, how you see this will generate a project with the next structure

Reviewer 3: Maybe the first occurrence of rsgem could be a link to its repo, so we can gain some visitors and contribute to the OSS OKR. What do you think?

> The first step to start is to install the rsgem, this is able to do easily by running `$ gem install rsgem`. After we have the gem installed, we have to run `$ rsgem new [name]`, with the name that we want for our gem. In particular, in this blog, we will be running the command `$ rsgem new blog_gem`, and this will generate a project with the following structure:

Reviewer: ```suggestion
The first step here is to install the rsgem by running `$ gem install rsgem`. Once the gem is installed, we have to run `$ rsgem new [name]`, with the name that we want for our gem included. In this blog, we will pay close attention we will be running the command `$ rsgem new blog_gem`, as this will generate a project with the following structure:
```

<img src="images/rsgem_structure.png"
     alt="Blog Gem"
     width="500" height="250"
     align="middle"/>

> It is important to mention that the gem rsgem is based on the command 'bundle gem', said command is the one which generates the directories and the files which we see. Now, we will mention the different components which are generated for the gem that we want to do.

Reviewer 2: ```suggestion
It is important to mention that the gem rsgem is based on the command 'bundle gem', said command is the one that generates the directories and the files that we see. Now, we will mention the different components which are generated for the gem that we want to create.
```

> It's important to mention that the gem rsgem is based on the command 'bundle gem', said command is the one that generates the directories and the files that we see. Now, we will mention the different components which are generated for the gem that we want to create.

Reviewer: ```suggestion
It's important to mention that the gem rsgem is based on the command 'bundle gem', as it is the command that generates the directories and the files that are on display. Therefore, we will look at the different components which are generated for the gem that we want to create.
```

> ## Bin folder

Reviewer: ```suggestion
## Bin Folder
```

<img src="images/bin_folder.png"
     alt="Bin folder"
     width="500" height="250"
     align="middle"/>

> In the Bin folder, we can find the console and setup files. The important thing to highlight here is the utility of console file where we can see the next code:

Reviewer: ```suggestion
In the Bin folder, we will find the console and setup files. The important thing here to highlight the utility of the console file, where we can see the next code:
```

```ruby
#!/usr/bin/env ruby

require 'bundler/setup'
require 'blog_gem'

# You can add fixtures and/or initialization code here to make experimenting
# with your gem easier. You can also use a different console, if you like.

# (If you use this, don't forget to add pry to your Gemfile!)
# require "pry"
# Pry.start

require 'irb'
IRB.start(__FILE__)
```

If we have to do test from the terminal with our gem, we only need the command bundle console 'require_name' to be able to access to the interactive ruby shell with each of our methods and dependecies of the gem. In our example of gem, we have just to do bundle console blog_gem.

> ## Lib folder

Reviewer: ```suggestion
## Lib Folder
```

<img src="images/lib_folder.png"
     alt="Lib folder"
     width="500" height="250"
     align="middle"/>

> In the Lib folder is where the magic will be done, here we will put our classes, our code, our methods that will be offered to those developers who want to use the gem that we make. If we look closely, we will see a folder with the name of our gem and if we look inside the folder, we will find a file with the name version. This version file is used to indicate what is the version of our gem and it is the file that is used to complete the version attribute of our gemspec file which we will see later.

Reviewer: ```suggestion
The Lib folder is where the magic happens, as it is where we put our classes, codes, and methods that will be on offer to the developers who want to use our gem. If you look closely, you will see a folder with the name of our gem, and if you look inside the folder, you will find a file with the name version. This version file is used to indicate the current version of our gem, and will also be used to complete the version attribute of our gemspec file which we will see later.
```

> On the other hand we have the file name_gem.rb (in our example 'blog_gem.rb'), which is the main file of the gem. This file contains code similar to the following:

Reviewer: ```suggestion
As well as this, we have the file name_gem.rb (in our example 'blog_gem.rb'), which is the main file of the gem. This file contains specific code similar to the following:
```

```ruby
require 'blog_gem/version'

module BlogGem
  class Error < StandardError; end
  # Your code goes here...
end

```

> As we can see, it is a class where, in principle, you could put code directly. That decision is up to the developer's wish, but generally no code is added to this file, instead of it, 'requires' are done to the files which contain the classes which offer the methods of our gem. We can also see this class does generic error handling which follows the ruby guidelines.

Reviewer: ```suggestion
As we can see here, it is a class wherein principle, you could put the code directly in. That decision is up to the developer, but generally, there won't be code added to this file. Instead of this, 'requires' will be added to the files, which contain the classes which offer the methods of our gem. We can also see this class covers generic error handling, which follows the ruby guidelines.
```

> ## Spec folder

Reviewer: ```suggestion
## Spec Folder
```

<img src="images/spec_folder.png"
     alt="Spec folder"
     width="500" height="250"
     align="middle"/>

> Both in Rootstrap and in the software world it is considered that good software always has to have unit tests which support the implemented code, for this very reason we use rspec. In the spec folder, we will put all the test files, where the unit tests of our gem will be implemented and which we can execute from the console with the rspec command.

Reviewer: ```suggestion
In the world of software development, it's understood that good software always has to have unit tests that support the implemented code. For this very reason, we use RSpec. Our test files must be put in the spec folder, as this will allow us to implement the unit tests of our gem, which we can execute from the console with the RSpec command.
```

> ## Gemspec file

Reviewer: ```suggestion
## Gemspec File
```

  spec.version       = BlogGem::VERSION
  spec.authors       = ['TimoPeraza']
  spec.email         = ['timothy@rootstrap.com']

  spec.summary       = ''
  spec.description   = ''
  spec.homepage      = ''
  spec.license       = 'MIT'
  spec.required_ruby_version = Gem::Requirement.new('>= 2.3.0')

  # spec.metadata["allowed_push_host"] = "TODO: Set to 'http://mygemserver.com'"

  # spec.metadata["homepage_uri"] = spec.homepage
  # spec.metadata["source_code_uri"] = "TODO: Put your gem's public repo URL here."
  # spec.metadata["changelog_uri"] = "TODO: Put your gem's CHANGELOG.md URL here."

  # Specify which files should be added to the gem when it is released.
  # The `git ls-files -z` loads the files in the RubyGem that have been added into git.
  spec.files = Dir['LICENSE.txt', 'README.md', 'lib/**/*']
  spec.bindir        = 'exe'
  spec.executables   = spec.files.grep(%r{^exe/}) { |f| File.basename(f) }
  spec.require_paths = ['lib']
  spec.add_development_dependency 'rake'
  spec.add_development_dependency 'reek'
  spec.add_development_dependency 'rspec'
  spec.add_development_dependency 'rubocop'
  spec.add_development_dependency 'simplecov', '~> 0.17.1'
end
```

> In this file we will put all the information about our gem, who made it, contact email, description, version, the files that compose the gem to do a release and also add all the external dependencies which the gem needs.

Reviewer: ```suggestion
In this file, we will put all the information for our gem i.e. who made it, contact email, description, and version. This file will also contain the files that compose the gem to do a release and the external dependencies which the gem needs.
```

> By default, rsgem adds the dev dependencies: rake, reek, rspec, rubocop, and simplecov. This is because in Rootstrap we focus a lot on the quality of the code, it seems to us something very important when developing therefore, we add these dependencies to control that our gem has good quality.

Reviewer: ```suggestion
By default, Rsgem adds the following dev dependencies: Rake, Reek, RSpec, Rubocop, and Simplecov. We do this because at Rootrsap we like to focus on the qualify of the code, and as a result, we add these dependencies to ensure our gem has the highest quality possible.
```

> ## Readme file

Reviewer: ```suggestion
## Readme File
```

> Finally, in the readme file we will add all the information about our gem, how it works, how to install it, the methods it offers and everything that is useful so another developer can use it.

Reviewer: ```suggestion
Finally, in the Readme file, we will need to add all of the information associated with our gem, for example, how it works, how to install it, the methods it offers, as well as anything that could be useful for a developer to use. 
```

> Now we have a general idea of the components that the project has, just get down to work and make your first gem! For more information about how to make a gem I strongly recommend looking at the page of [rubygems](https://guides.rubygems.org/make-your-own-gem/)

Reviewer: ```suggestion
Now that we have a general idea of the components that a project has, you are ready to get down to work and develop your first gem! For more information about how to do this, I strongly recommend looking at the page of [rubygems](https://guides.rubygems.org/make-your-own-gem/)
```
